### DISCUSSION TITLE
Copy form from one pdf to another

### DISCUSSION THREAD

**Discussion by cmyers009** [pymupdf/PyMuPDF#3702] [Looking for help]
Sometimes a pdf is a form pdf which has widgets.

I want to be able to copy the widgets from one PDF to another.

I am able to do this successfully for certain types of widgets (type 1, 2, 3, 4, and most of the time 7) but my method does not work for radio groups (type 5)

Why does it not work for these? Is there something I can change to make this work?

Is there any risks with my current implementation, if I was to support types 1, 2, and 7 but not support type 5.

Lastly, I found that sometimes PDF forms of type '7' (text) may have images embedded in the text.  Copying the form text field over this way does not seem to work, and the image gets deleted.

[radios_input.pdf](https://github.com/user-attachments/files/16289420/radios_input.pdf)
[radios_output.pdf](https://github.com/user-attachments/files/16289438/radios_output.pdf)

Here is an example of a PDF that works (does not contain type 5)

[Widget_example_2.pdf](https://github.com/user-attachments/files/16289536/Widget_example_2.pdf)

`ValueError: bad xref
Traceback (most recent call last):
  File "c:\users\XXX\desktop\test_pdf\copy_widgets.py", line 32, in copy_pdf_with_forms
    dst_widget = dst_page.add_widget(src_widget)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XXX\anaconda3\Lib\site-packages\pymupdf\__init__.py", line 8448, in add_widget
    widget.update()
  File "C:\Users\XXX\anaconda3\Lib\site-packages\pymupdf\__init__.py", line 7142, in update
    self._validate()
  File "C:\Users\XXX\anaconda3\Lib\site-packages\pymupdf\__init__.py", line 7056, in _validate
    self._checker()  # any field_type specific checks
    ^^^^^^^^^^^^^^^
  File "C:\Users\XXX\anaconda3\Lib\site-packages\pymupdf\__init__.py", line 6947, in _checker
    kids_type, kids_value = doc.xref_get_key(self.xref, "Parent/Kids")
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XXX\anaconda3\Lib\site-packages\pymupdf\__init__.py", line 5787, in xref_get_key
    raise ValueError( MSG_BAD_XREF)
ValueError: bad xref`

```python
import fitz
import json
import traceback
def copy_pdf_with_forms(input_path, output_path):
    src_doc = fitz.open(input_path)
    dst_doc = fitz.open()
    
    # List of attributes to copy
    attributes_to_copy = [
        'field_name', 'field_flags', 'field_type', 'field_value',
        'text_font', 'text_fontsize', 'text_color', 'fill_color', 'border_color',
        'choice_values'#, 'rect'
    ]

    
    radio_groups = {}
    
    for i, src_page in enumerate(src_doc):
        #dst_page = #dst_doc.new_page(width=src_page.rect.width, height=src_page.rect.height)
        
        dst_doc.insert_pdf(src_doc, from_page=i, to_page=i)
        dst_page=dst_doc[-1]
        
        for src_widget in src_page.widgets():
            try:
                if src_widget.field_type == 5:
                    print(src_widget.field_type_string)
                    print("WARNING: radio button type not supported.")
                    continue
                
                # Create a new widget
                dst_widget = dst_page.add_widget(src_widget)
                
                # Store the parent page reference directly
                dst_widget.parent_page = dst_page
                
                # Copy attributes
                for attr in attributes_to_copy:
                    if hasattr(src_widget, attr):
                        try:
                            setattr(dst_widget, attr, getattr(src_widget, attr))
                        except AttributeError:
                            traceback.print_exc()
                            print(f"Warning: Could not set attribute {attr} for {src_widget.field_name}")
                
                # Set the rectangle using set_rect method
                try:
                    dst_widget.set_rect(src_widget.rect)
                except AttributeError:
                    traceback.print_exc()
                    print(f"Warning: Could not set rect for {src_widget.field_name}")
                
                
                print(f"Copied widget: {src_widget.field_name}, Type: {src_widget.field_type}, Value: '{src_widget.field_value}'")
            except Exception as e:
                traceback.print_exc()
                print(f"Warning: Could not copy widget {src_widget.field_name}. Error: {str(e)}")
    

    dst_doc.save(output_path)
    src_doc.close()
    dst_doc.close()

def print_form_fields(pdf_path):
    doc = fitz.open(pdf_path)
    form_fields = []
    for page in doc:
        for widget in page.widgets():
            field_info = {
                "name": widget.field_name,
                "type": widget.field_type,
                "value": widget.field_value,
                "rect": [round(coord, 2) for coord in widget.rect],
                "flags": widget.field_flags
            }
            form_fields.append(field_info)
    doc.close()
    
    json_path = pdf_path.replace('.pdf', ".json")
    with open(json_path, "w") as f:
        json.dump(form_fields, f, indent=2)
    print(f"Form fields saved to {json_path}")

# Usage
input_path = "radios.pdf"
#input_path = "input_with_forms.pdf"
#input_path = "forms_test.pdf"
#input_path = "Widget_example_2.pdf"
output_path = "output_with_forms.pdf"
copy_pdf_with_forms(input_path, output_path)
print_form_fields(input_path)
print_form_fields(output_path)

**Comment by JorjMcKie** [comment]
I would not rely on this working ever. Radio buttons usually come in groups with a mechanism that automatically switches off all other radio buttons in the group when one button is switched on.
While PyMuPDF allows inserting single radio buttons, the creation of such groups is not.
Also specialties as you mention them (image displays within text fields etc.) are not supported.

**Reply by cmyers009** [reply]
Thanks for the advice. 

Instead of trying to move the forms from one PDF to another, is there any other approach that may achieve similar results. 

My goal is to modify the pdf to add some enhancements, like rotating certain pages, adding OCR, removing blank pages. 

The way I am doing this is starting with the source pdf and creating a new PDF with the pixmap, and an OCR overlay. 

If instead of creating a new pdf I was copying things over to, if I modified my existing PDF in place, could that work? 

Or is there any way to copy the entire PDF page (including form fields) if I detect that the page contains a form?

**Reply by JorjMcKie** [reply]
You can "bake in" annotations and / or fields. This entails converting these objects to **_permanent_** page content: neither fields nor annotations are part of a page's appearance source, but floating "above" that - comparable to dust on the glass cover of a picture on a wall.
`Document.bake()` changes that.
Things will look alike but there will be not annotations or fields anymore.

After doing this, a source page can be copied to a target PDF without worries via e.g. `target.insert_pdf(source, from_page=n, to_page=m)`.
